File: 0589-n-ary-tree-preorder-traversal/0589-n-ary-tree-preorder-traversal.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node|null} root
 * @return {number[]}
 */

var preorder = function(root) {
    var stack = [];
    var output = [];
    
    if (!root) {
        return output;
    }
    
    stack.push(root);
    
    while (stack.length > 0) {
        var node = stack.pop();
        
        output.push(node.val);
        
        node.children.reverse();
        node.children.forEach(child => {
            stack.push(child);
        });
    }
    
    return output;
};

File: 0589-n-ary-tree-preorder-traversal/0589-n-ary-tree-preorder-traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> preorder(Node root) {
        LinkedList<Node> stack = new LinkedList();
        LinkedList<Integer> output = new LinkedList();
        
        if (root == null) {
            return output;
        }
        
        stack.add(root);
        while (!stack.isEmpty()){
            Node next = stack.pollLast();
            output.add(next.val);
            
            Collections.reverse(next.children);
            for (Node child : next.children) {
                stack.add(child);
            }
        }
        
        return output;
    }
}